### ISSUE TITLE
Perhaps the CARN model is incorrectly implemented?

### ISSUE THREAD

**Issue by wangqiim** [Xiangtaokong/ClassSR#42]
https://github.com/Xiangtaokong/ClassSR/blob/0ebefe75b2d5d08cffd8613c78d6dbb323dbd2c5/codes/models/archs/arch_util.py#L103-L114
I face train problems when train class_carn, finally I located the problem in above code.
In CARN official repo, I found there are no bn layer, So may be the CARN model is incorrectly implemented?

https://github.com/nmhkahn/CARN-pytorch/blob/master/carn/model/ops.py#L33-L48

**Comment by Xiangtaokong** [comment]
Yeah, we do incorrectly use CARN with BN. (official CARN has no BN). But this operation does not affect the training, and it does not affect the conclusion of ClassSR. (''CARN with BN'' can be regraded as a new accelerated network)

**Comment by wangqiim** [comment]
> 

Yeah. But If add BN, I realize when train classsr, the pre train branch will crash? Because the BN params can't be freeze, So if the module.trainning is True, network forward will always update BN params(not base on gradient), and the loss will increase dramatically compare with pre train branch.

**Comment by wangqiim** [comment]
> Yeah, we do incorrectly use CARN with BN. (official CARN has no BN). But this operation does not affect the training, and it does not affect the conclusion of ClassSR. (''CARN with BN'' can be regraded as a new accelerated network)

It may eventually converges, But I think it  probably is a small issue :)

**Comment by Xiangtaokong** [comment]
> > @Xiangtaokong
> 
> Yeah. But If add BN, I realize when train classsr, the pre train branch will crash?(SR module loss increase)Then I remove bn, then it didn't happen. Because the BN params can't be freeze, So if the module.trainning is True, network forward will always update BN params(not base on gradient), and the loss will increase dramatically compare with pre train branch.

We've seen this before, and the classifier converged.  I remember that I have been debugging with my collaborator for a long time and BN can be frozen in the current code by settiing v.equires_grad =False instead of module.trainning.  You can freeze BN directly by setting fix_SR_module: true in yml. And, of course, you can just not use BN.

https://github.com/Xiangtaokong/ClassSR/blob/0ebefe75b2d5d08cffd8613c78d6dbb323dbd2c5/codes/models/ClassSR_model.py#L77

**Comment by wangqiim** [comment]
> > > @Xiangtaokong
> > 
> > 
> > Yeah. But If add BN, I realize when train classsr, the pre train branch will crash?(SR module loss increase)Then I remove bn, then it didn't happen. Because the BN params can't be freeze, So if the module.trainning is True, network forward will always update BN params(not base on gradient), and the loss will increase dramatically compare with pre train branch.
> 
> We've seen this before, and the classifier converged. I remember that I have been debugging with my collaborator for a long time and BN can be frozen in the current code by settiing v.equires_grad =False instead of module.trainning. You can freeze BN directly by setting fix_SR_module: true in yml. And, of course, you can just not use BN.
> 
> https://github.com/Xiangtaokong/ClassSR/blob/0ebefe75b2d5d08cffd8613c78d6dbb323dbd2c5/codes/models/ClassSR_model.py#L77

Make sense. Thank u. I had set 'fix_SR_module' and try to add detach() before. And I "step debug " or print log frequently, the intermediate results confuse me. I think it is about design of BN, There are some static params and runtime params in BN,  but the runtime params can only be freezed when set module.eval(). (require_grad or detach just only prevent gradient propagation). These affect the intermediate result during training I guess:)
(https://discuss.pytorch.org/t/batchnorm-with-requires-grad-false-and-training-true-in-validation/58986)

**Comment by Xiangtaokong** [comment]
> > > > @Xiangtaokong
> > > 
> > > 
> > > Yeah. But If add BN, I realize when train classsr, the pre train branch will crash?(SR module loss increase)Then I remove bn, then it didn't happen. Because the BN params can't be freeze, So if the module.trainning is True, network forward will always update BN params(not base on gradient), and the loss will increase dramatically compare with pre train branch.
> > 
> > 
> > We've seen this before, and the classifier converged. I remember that I have been debugging with my collaborator for a long time and BN can be frozen in the current code by settiing v.equires_grad =False instead of module.trainning. You can freeze BN directly by setting fix_SR_module: true in yml. And, of course, you can just not use BN.
> > https://github.com/Xiangtaokong/ClassSR/blob/0ebefe75b2d5d08cffd8613c78d6dbb323dbd2c5/codes/models/ClassSR_model.py#L77
> 
> Make sense. Thank u. I had set 'fix_SR_module' and try to add detach() before. And I "step debug " or print log frequently, the intermediate results confuse me. I think it is about design of BN, There are some static params and runtime params in BN, but the runtime params can only be freezed when set module.eval(). (require_grad or detach just only prevent gradient propagation). These affect the intermediate result during training I guess:) (https://discuss.pytorch.org/t/batchnorm-with-requires-grad-false-and-training-true-in-validation/58986)

I don't know, when we saw that this would keep the output of BN unchanged, we thought it was frozen, maybe there's something else that needs to be frozen ……
